File: .github/scripts/fetch_external.py
"""
Fetch external content from Confluence or Jira.

Usage:
    python fetch_external.py confluence <page_id> [--debug]
    python fetch_external.py jira <issue_key> [--debug]

Environment Variables (required):
    CONFLUENCE_USERNAME: Your Confluence username
    CONFLUENCE_TOKEN: Your Confluence API token
    JIRA_USERNAME: Your Jira username  
    JIRA_TOKEN: Your Jira API token

Create a .env file in the same directory with these variables.
"""

import sys
import os
import requests
import json
import argparse
from urllib.parse import urlparse
from dotenv import load_dotenv

# Load environment variables from .env file if it exists
load_dotenv()

def fetch_confluence_page(url, auth):
    """Fetch Confluence page content"""
    # Handle short links by following redirects
    try:
        redirect_response = requests.head(url, auth=auth, allow_redirects=True)
        final_url = redirect_response.url
        print(f"Redirected to: {final_url}")
    except:
        final_url = url

    # Extract page ID from URL
    parsed = urlparse(final_url)
    path_parts = parsed.path.split('/')
    page_id = None
    for i, part in enumerate(path_parts):
        if part == 'pages' and i + 1 < len(path_parts):
            page_id = path_parts[i + 1]
            break

    if not page_id:
        return {"error": "Could not extract page ID from Confluence URL"}

    # Try API v2 first (recommended)
    api_url = f"{parsed.scheme}://{parsed.netloc}/wiki/api/v2/pages/{page_id}?body-format=ATLAS_DOC_FORMAT"
    attachments_url = f"{parsed.scheme}://{parsed.netloc}/wiki/api/v2/pages/{page_id}/attachments"

    try:
        # Fetch page info with content
        response = requests.get(api_url, auth=auth)
        response.raise_for_status()
        page_data = response.json()

        # Fetch attachments
        attachments_response = requests.get(attachments_url, auth=auth)
        attachments = []
        if attachments_response.status_code == 200:
            attachments_data = attachments_response.json()
            attachments = [
                {
                    "id": att.get("id"),
                    "title": att.get("title"),
                    "fileName": att.get("fileName"),
                    "fileSize": att.get("fileSize"),
                    "mediaType": att.get("mediaType"),
                    "downloadUrl": f"{parsed.scheme}://{parsed.netloc}/wiki/download/attachments/{page_id}/{att.get('fileName')}"
                }
                for att in attachments_data.get("results", [])
            ]

        return {
            "title": page_data.get("title", ""),
            "content": page_data.get("body", {}).get("atlas_doc_format", {}).get("value", ""),
            "attachments": attachments,
            "url": url,
            "type": "confluence"
        }
    except Exception as e:
        # Fallback to API v1 if v2 fails
        try:
            api_url_v1 = f"{parsed.scheme}://{parsed.netloc}/wiki/rest/api/content/{page_id}?expand=body.storage"
            attachments_url_v1 = f"{parsed.scheme}://{parsed.netloc}/wiki/rest/api/content/{page_id}/child/attachment"

            response = requests.get(api_url_v1, auth=auth)
            response.raise_for_status()
            data = response.json()

            # Fetch attachments v1
            attachments_response = requests.get(attachments_url_v1, auth=auth)
            attachments = []
            if attachments_response.status_code == 200:
                attachments_data = attachments_response.json()
                attachments = [
                    {
                        "id": att.get("id"),
                        "title": att.get("title"),
                        "fileName": att.get("fileName") or att.get("title"),
                        "fileSize": att.get("extensions", {}).get("fileSize"),
                        "mediaType": att.get("extensions", {}).get("mediaType"),
                        "downloadUrl": att.get("_links", {}).get("download", "")
                    }
                    for att in attachments_data.get("results", [])
                ]

            return {
                "title": data.get("title", ""),
                "content": data.get("body", {}).get("storage", {}).get("value", ""),
                "attachments": attachments,
                "url": url,
                "type": "confluence"
            }
        except Exception as e2:
            return {"error": f"Failed to fetch Confluence page: {str(e2)}"}

def fetch_jira_issue(url, auth):
    """Fetch Jira issue details"""
    parsed = urlparse(url)
    path_parts = parsed.path.split('/')
    issue_key = None
    for part in path_parts:
        if '-' in part and part.split('-')[0].isalpha():
            issue_key = part
            break

    if not issue_key:
        return {"error": "Could not extract issue key from Jira URL"}

    api_url = f"{parsed.scheme}://{parsed.netloc}/rest/api/2/issue/{issue_key}"

    try:
        response = requests.get(api_url, auth=auth)
        response.raise_for_status()
        data = response.json()

        return {
            "key": data.get("key", ""),
            "title": data.get("fields", {}).get("summary", ""),
            "description": data.get("fields", {}).get("description", ""),
            "status": data.get("fields", {}).get("status", {}).get("name", ""),
            "assignee": data.get("fields", {}).get("assignee", {}).get("displayName", "") if data.get("fields", {}).get("assignee") else None,
            "url": url,
            "type": "jira"
        }
    except Exception as e:
        return {"error": f"Failed to fetch Jira issue: {str(e)}"}

def main():
    parser = argparse.ArgumentParser(description='Fetch content from Confluence or Jira')
    parser.add_argument('url', help='URL to fetch')
    parser.add_argument('--type', choices=['confluence', 'jira'], help='Type of source (auto-detected if not specified)')

    args = parser.parse_args()

    # Determine type from URL if not specified
    if not args.type:
        if 'atlassian.net' in args.url and ('wiki' in args.url or 'confluence' in args.url):
            args.type = 'confluence'
        elif 'atlassian.net' in args.url or 'jira' in args.url:
            args.type = 'jira'
        else:
            print(json.dumps({"error": "Could not determine source type from URL"}))
            sys.exit(1)

    # Get auth from environment variables only
    username = os.getenv(f"{args.type.upper()}_USERNAME")
    token = os.getenv(f"{args.type.upper()}_TOKEN")
    
    print(f"Using authentication for {args.type}: username={username}, token={'***' if token else None}")
    
    if token:
        auth = (username or '', token)
    else:
        print(json.dumps({"error": "Authentication required. Set environment variables in .env file"}))
        sys.exit(1)

    # Fetch content
    if args.type == 'confluence':
        result = fetch_confluence_page(args.url, auth)
    elif args.type == 'jira':
        result = fetch_jira_issue(args.url, auth)
    else:
        result = {"error": "Unsupported type"}

    print(json.dumps(result, indent=2))

if __name__ == '__main__':
    main()

File: .github/scripts/fetch_external.mjs
#!/usr/bin/env node
/**
 * Fetch external content from Confluence or Jira.
 * 
 * Usage:
 *   node fetch_external.mjs <url> --type confluence|jira
 *   node fetch_external.mjs "https://xxx.atlassian.net/wiki/x/abc" --type confluence
 *   node fetch_external.mjs "https://xxx.atlassian.net/browse/PROJ-123" --type jira
 * 
 * Environment Variables (in .env file):
 *   CONFLUENCE_USERNAME: Your Confluence email
 *   CONFLUENCE_TOKEN: Your Confluence API token
 *   JIRA_USERNAME: Your Jira email
 *   JIRA_TOKEN: Your Jira API token
 */

import { readFileSync, existsSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Load .env file
function loadEnv() {
  const envPath = join(__dirname, '.env');
  if (!existsSync(envPath)) {
    console.error('Error: .env file not found at', envPath);
    console.error('Run: cp .github/scripts/env.example .github/scripts/.env');
    process.exit(1);
  }
  
  const envContent = readFileSync(envPath, 'utf-8');
  const env = {};
  
  envContent.split('\n').forEach(line => {
    const trimmed = line.trim();
    if (trimmed && !trimmed.startsWith('#')) {
      const [key, ...valueParts] = trimmed.split('=');
      if (key) {
        env[key.trim()] = valueParts.join('=').trim().replace(/^["']|["']$/g, '');
      }
    }
  });
  
  return env;
}

// Parse command line arguments
function parseArgs() {
  const args = process.argv.slice(2);
  const result = { url: null, type: null };
  
  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--type' && args[i + 1]) {
      result.type = args[i + 1];
      i++;
    } else if (!args[i].startsWith('--')) {
      result.url = args[i];
    }
  }
  
  return result;
}

// Detect type from URL
function detectType(url) {
  if (url.includes('atlassian.net') && (url.includes('/wiki') || url.includes('confluence'))) {
    return 'confluence';
  }
  if (url.includes('atlassian.net') && url.includes('/browse/')) {
    return 'jira';
  }
  return null;
}

// Create Basic Auth header
function createAuthHeader(username, token) {
  const credentials = Buffer.from(`${username}:${token}`).toString('base64');
  return `Basic ${credentials}`;
}

// Fetch with error handling
async function fetchWithAuth(url, auth) {
  const response = await fetch(url, {
    headers: {
      'Authorization': auth,
      'Accept': 'application/json',
      'Content-Type': 'application/json'
    },
    redirect: 'follow'
  });
  
  if (!response.ok) {
    throw new Error(`HTTP ${response.status}: ${response.statusText}`);
  }
  
  return response.json();
}

// Follow redirects to get final URL
async function getRedirectedUrl(url, auth) {
  try {
    const response = await fetch(url, {
      method: 'HEAD',
      headers: { 'Authorization': auth },
      redirect: 'follow'
    });
    return response.url;
  } catch {
    return url;
  }
}

// Extract page ID from Confluence URL
function extractPageId(url) {
  const parsed = new URL(url);
  const pathParts = parsed.pathname.split('/');
  
  for (let i = 0; i < pathParts.length; i++) {
    if (pathParts[i] === 'pages' && pathParts[i + 1]) {
      return pathParts[i + 1];
    }
  }
  
  return null;
}

// Extract issue key from Jira URL
function extractIssueKey(url) {
  const parsed = new URL(url);
  const pathParts = parsed.pathname.split('/');
  
  for (const part of pathParts) {
    if (part.includes('-') && /^[A-Z]+-\d+$/i.test(part)) {
      return part.toUpperCase();
    }
  }
  
  return null;
}

// Fetch Confluence page
async function fetchConfluencePage(url, auth) {
  console.error(`Fetching Confluence page: ${url}`);
  
  // Follow redirects for short links
  const finalUrl = await getRedirectedUrl(url, auth);
  console.error(`Redirected to: ${finalUrl}`);
  
  const pageId = extractPageId(finalUrl);
  if (!pageId) {
    return { error: 'Could not extract page ID from Confluence URL' };
  }
  
  const parsed = new URL(finalUrl);
  const baseUrl = `${parsed.protocol}//${parsed.host}`;
  
  try {
    // Try API v2 first
    const apiUrl = `${baseUrl}/wiki/api/v2/pages/${pageId}?body-format=ATLAS_DOC_FORMAT`;
    const attachmentsUrl = `${baseUrl}/wiki/api/v2/pages/${pageId}/attachments`;
    
    const pageData = await fetchWithAuth(apiUrl, auth);
    
    // Fetch attachments
    let attachments = [];
    try {
      const attachmentsData = await fetchWithAuth(attachmentsUrl, auth);
      attachments = (attachmentsData.results || []).map(att => ({
        id: att.id,
        title: att.title,
        fileName: att.fileName,
        fileSize: att.fileSize,
        mediaType: att.mediaType,
        downloadUrl: `${baseUrl}/wiki/download/attachments/${pageId}/${att.fileName}`
      }));
    } catch {
      // Attachments fetch failed, continue without
    }
    
    return {
      title: pageData.title || '',
      content: pageData.body?.atlas_doc_format?.value || '',
      attachments,
      url,
      type: 'confluence'
    };
    
  } catch (e) {
    // Fallback to API v1
    console.error(`API v2 failed, trying v1: ${e.message}`);
    
    try {
      const apiUrlV1 = `${baseUrl}/wiki/rest/api/content/${pageId}?expand=body.storage`;
      const attachmentsUrlV1 = `${baseUrl}/wiki/rest/api/content/${pageId}/child/attachment`;
      
      const data = await fetchWithAuth(apiUrlV1, auth);
      
      let attachments = [];
      try {
        const attachmentsData = await fetchWithAuth(attachmentsUrlV1, auth);
        attachments = (attachmentsData.results || []).map(att => ({
          id: att.id,
          title: att.title,
          fileName: att.fileName || att.title,
          fileSize: att.extensions?.fileSize,
          mediaType: att.extensions?.mediaType,
          downloadUrl: att._links?.download || ''
        }));
      } catch {
        // Attachments fetch failed
      }
      
      return {
        title: data.title || '',
        content: data.body?.storage?.value || '',
        attachments,
        url,
        type: 'confluence'
      };
      
    } catch (e2) {
      return { error: `Failed to fetch Confluence page: ${e2.message}` };
    }
  }
}

// Fetch Jira issue
async function fetchJiraIssue(url, auth) {
  console.error(`Fetching Jira issue: ${url}`);
  
  const issueKey = extractIssueKey(url);
  if (!issueKey) {
    return { error: 'Could not extract issue key from Jira URL' };
  }
  
  const parsed = new URL(url);
  const baseUrl = `${parsed.protocol}//${parsed.host}`;
  const apiUrl = `${baseUrl}/rest/api/2/issue/${issueKey}`;
  
  try {
    const data = await fetchWithAuth(apiUrl, auth);
    
    return {
      key: data.key || '',
      title: data.fields?.summary || '',
      description: data.fields?.description || '',
      status: data.fields?.status?.name || '',
      assignee: data.fields?.assignee?.displayName || null,
      url,
      type: 'jira'
    };
    
  } catch (e) {
    return { error: `Failed to fetch Jira issue: ${e.message}` };
  }
}

// Main
async function main() {
  const args = parseArgs();
  
  if (!args.url) {
    console.error('Usage: node fetch_external.mjs <url> --type confluence|jira');
    process.exit(1);
  }
  
  // Auto-detect type if not specified
  const type = args.type || detectType(args.url);
  if (!type) {
    console.error(JSON.stringify({ error: 'Could not determine source type from URL. Use --type' }));
    process.exit(1);
  }
  
  // Load environment
  const env = loadEnv();
  
  const username = env[`${type.toUpperCase()}_USERNAME`];
  const token = env[`${type.toUpperCase()}_TOKEN`];
  
  console.error(`Using authentication for ${type}: username=${username}, token=${token ? '***' : 'NOT SET'}`);
  
  if (!token) {
    console.error(JSON.stringify({ 
      error: 'Authentication required. Set environment variables in .env file',
      required: [`${type.toUpperCase()}_USERNAME`, `${type.toUpperCase()}_TOKEN`]
    }));
    process.exit(1);
  }
  
  const auth = createAuthHeader(username || '', token);
  
  let result;
  if (type === 'confluence') {
    result = await fetchConfluencePage(args.url, auth);
  } else if (type === 'jira') {
    result = await fetchJiraIssue(args.url, auth);
  } else {
    result = { error: 'Unsupported type' };
  }
  
  console.log(JSON.stringify(result, null, 2));
}

main().catch(e => {
  console.error(JSON.stringify({ error: e.message }));
  process.exit(1);
});
